### PR TITLE
Add IPv6

### PR DESCRIPTION
**Pull request by NicolasDP**
Another PR will follow to update the parser and the IPv4 parser utilising `Foundation.Parser`.

### CODE REVIEW

File: Foundation/Network/IPv6.hs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+import Foundation.Primitive
+import Foundation.Collection (intercalate, span)
+import Foundation.String (String)
+import Foundation.Bits
+
+-- | IPv6 data type
+data IPv6 = IPv6 {-# UNPACK #-} !Word64 {-# UNPACK #-} !Word64
+  deriving (Eq, Ord, Typeable)
+instance Hashable IPv6 where
+    hashMix (IPv6 w1 w2) = hashMix (w1, w2)
```
Comment (vincenthz) [inline Foundation/Network/IPv6.hs]: hashMix w2 . hashMix w1

File: Foundation/Network/IPv6.hs [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+instance Show IPv6 where
+    show = toLString
+instance IsString IPv6 where
+    fromString = fromLString
+instance Storable IPv6 where
+    peek ptr = fromTuple <$>
+        (   (,,,,,,,)
+        <$> (fromBE <$> peekOff ptr' 1)
```
Comment (vincenthz) [inline Foundation/Network/IPv6.hs]: start at 1 ?

File: Foundation/Network/IPv6.hs [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        <*> (fromBE <$> peekOff ptr' 6)
+        <*> (fromBE <$> peekOff ptr' 7)
+        <*> (fromBE <$> peekOff ptr' 8)
+        )
+      where
+        ptr' :: Ptr (BE Word16)
+        ptr' = castPtr ptr
+    poke ptr ipv6 = do
+        let (i1,i2,i3,i4,i5,i6,i7,i8) = toTuple ipv6
+         in pokeOff ptr' 1 (toBE i1)
```
Comment (vincenthz) [inline Foundation/Network/IPv6.hs]: start at 1 ?

File: Foundation/Network/IPv6.hs [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+-- | create an IPv6 from the given tuple
+fromTuple :: (Word16, Word16, Word16, Word16, Word16, Word16, Word16, Word16)
+          -> IPv6
+fromTuple (i1, i2, i3, i4, i5, i6, i7, i8) = IPv6 hi low
+  where
+    f :: Word16 -> Word64
+    f = fromIntegral
+    hi, low :: Word64
+    hi =    (f i1 .<<. 48) .&. 0xFFFF000000000000
```
Comment (vincenthz) [inline Foundation/Network/IPv6.hs]: there's no need to mask after shifting, a Word16 is guarantee to be in the mask `0xFFFF`

File: Foundation/Network/IPv6.hs [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+-- | decompose an IPv6 into a tuple
+toTuple :: IPv6 -> (Word16,Word16,Word16,Word16,Word16,Word16,Word16,Word16)
+toTuple (IPv6 hi low) =
+    (f w1, f w2, f w3, f w4, f w5, f w6, f w7, f w8)
+  where
+    f :: Word64 -> Word16
+    f = fromIntegral
+    w1, w2, w3, w4, w5, w6, w7, w8 :: Word64
+    w1 = hi  .>>. 48 .&. 0x0000FFFF
```
Comment (vincenthz) [inline Foundation/Network/IPv6.hs]: fromIntegral is guarantee to mask to `Word16` (`0x0000FFFF`), so this is redundant

File: Foundation/Network/IPv6.hs (hunk 1 above)
Comment (NicolasDP) [inline Foundation/Network/IPv6.hs]: thanks, this has been fixed now

File: Foundation/Network/IPv6.hs (hunk 2 above)
Comment (NicolasDP) [inline Foundation/Network/IPv6.hs]: OMG... good spot.

**Pull request merged**